File: src/lib/core/invoice_message.dart
import 'package:intl/intl.dart';

import '../data/models/invoice.dart';
import 'supabase_client.dart';

/// Link ẢNH HOÁ ĐƠN của 1 hoá đơn — Edge Function `i`.
///
/// Tên hàm chỉ một chữ `i` là **cố ý**, không phải viết tắt cho gọn: xem giải
/// thích ngân sách ký tự ở [buildInvoiceSmsMessage] ngay bên dưới.
///
/// Dạng **đường dẫn sạch**, KHÔNG dùng `?code=...`: test thật 17/09/2026 cho
/// thấy nhà mạng chặn tin nhắn chứa `?` hoặc `&` (xem `docs/SMS-HOA-DON.md`
/// mục 6.4). Bỏ luôn `https://` cho ngắn — điện thoại vẫn nhận ra là link.
String invoiceImageLink(Invoice invoice) {
  final host = SupabaseConfig.url.replaceFirst(RegExp(r'^https?://'), '');
  return '$host/functions/v1/i/${invoice.publicCode}';
}

/// Nội dung SMS gửi Tenant cho 1 hoá đơn. Dùng chung cho B-05 (gửi đơn lẻ) và
/// B-03 (gửi hàng loạt, qua bản sinh đôi ở `_shared/invoice_message.ts`).
///
/// ⚠️ **Câu chữ bị khoá theo mẫu của brandname đi mượn, không tự viết được.**
/// eSMS chỉ cho gửi đúng những mẫu nội dung đã được nhà mạng duyệt cho từng
/// brandname; sai một dấu chấm là bị trả `CodeResult 146 — Sai template`
/// (kiểm chứng thật 18/09: đổi "Tran trong." thành "Tran trong!" là bị chặn,
/// và không bị trừ tiền). Dự án chưa có brandname riêng nên đang mượn
/// brandname demo `Baotrixemay` của eSMS.
///
/// ## Ngân sách ký tự — vì sao nội dung trông cụt lủn như vậy
///
/// Một đoạn SMS bảng mã GSM-7 chứa **160 ký tự**; vượt một ký tự là thành 2
/// đoạn, **trả tiền gấp đôi** (đo thật 18/09: 1.640đ cho 2 đoạn, ~820đ cho 1).
///
/// Riêng phần chữ CỐ ĐỊNH của mẫu mượn đã ăn **93 ký tự** ("Quy khach da den
/// thoi gian bao tri lan … xe … Vui long lien he … de duoc huong dan. Tran
/// trong.") — toàn chuyện bảo trì xe máy, vô nghĩa với người thuê, mà không
/// sửa được một chữ. Còn đúng **67 ký tự** cho mọi thứ có ích:
///
/// | Thành phần | Ký tự |
/// |---|---|
/// | Link ảnh hoá đơn | 60 |
/// | Kỳ (`MM/YY`) | 5 |
/// | Ô cuối (`BT`) | 2 |
/// | **Tổng cả tin** | **160 — vừa đúng 1 đoạn** |
///
/// Mọi thành phần đều dài CỐ ĐỊNH (tên miền Supabase 32, mã tra cứu luôn 12
/// ký tự) nên con số 160 là chắc chắn, không có hoá đơn nào vọt lên 2 đoạn.
///
/// **Số tiền và hạn thanh toán cố tình KHÔNG nằm trong tin nhắn** — dungtv
/// chốt 18/09/2026. Nhét chúng vào là tin dài 199 ký tự, thành 2 đoạn, đắt
/// gấp đôi. Người thuê bấm link là thấy đầy đủ hoá đơn kèm mã QR.
///
/// Khi mua được tên miền riêng (link còn ~17 ký tự) thì dư ra ~43 ký tự — lúc
/// đó nhét lại số tiền và hạn thanh toán vào mà vẫn giữ 1 đoạn.
String buildInvoiceSmsMessage(Invoice invoice) {
  final ky = DateFormat('MM/yy').format(invoice.periodStart); // 5
  final link = invoiceImageLink(invoice); // 60
  return 'Quy khach da den thoi gian bao tri lan $ky xe $link. '
      'Vui long lien he BT de duoc huong dan. Tran trong.';
}

File: supabase/functions/_shared/invoice_message.ts
// _shared/invoice_message.ts
//
// Dựng nội dung SMS/Zalo gửi Tenant cho 1 hoá đơn — bản TypeScript, ĐỒNG BỘ
// với `src/lib/core/invoice_message.dart` (Flutter, dùng ở B-05 gửi đơn lẻ).
// Dùng ở đây cho `generate-invoice` (mode "batchSend" — tạo + gửi hàng loạt
// chạy hẳn phía backend, xem docs/DECISIONS.md 2026-09-14).
//
// BR-NOTI-07: nội dung LUÔN tiếng Việt, không phụ thuộc ngôn ngữ UI.

const VN_BANKS: { bin: string; name: string }[] = [
  { bin: "970436", name: "Vietcombank" },
  { bin: "970415", name: "VietinBank" },
  { bin: "970418", name: "BIDV" },
  { bin: "970405", name: "Agribank" },
  { bin: "970407", name: "Techcombank" },
  { bin: "970416", name: "ACB" },
  { bin: "970422", name: "MB Bank" },
  { bin: "970432", name: "VPBank" },
  { bin: "970423", name: "TPBank" },
  { bin: "970403", name: "Sacombank" },
  { bin: "970437", name: "HDBank" },
  { bin: "970443", name: "SHB" },
  { bin: "970441", name: "VIB" },
  { bin: "970448", name: "OCB" },
  { bin: "970426", name: "MSB" },
  { bin: "970440", name: "SeABank" },
  { bin: "970431", name: "Eximbank" },
];

export function bankByBin(bin: string | null | undefined) {
  if (!bin) return null;
  return VN_BANKS.find((b) => b.bin === bin) ?? null;
}

function formatVndNumber(n: number): string {
  return Math.round(n).toLocaleString("en-US");
}

function formatDdMm(dateStr: string): string {
  const d = new Date(dateStr + "T00:00:00Z");
  const dd = String(d.getUTCDate()).padStart(2, "0");
  const mm = String(d.getUTCMonth() + 1).padStart(2, "0");
  return `${dd}/${mm}`;
}

/// Nội dung SMS hoá đơn gửi hàng loạt (B-03 → `generate-invoice` mode
/// "batchSend"). PHẢI giống hệt bản gửi đơn lẻ ở `src/lib/core/invoice_message.dart`
/// — cùng một mẫu, cùng một brandname.
///
/// Bản trước 18/09/2026 dựng một câu hoàn toàn khác ("BizTown Rent Manager:
/// Hoá đơn phòng ..."), hỏng hai đường cùng lúc:
///   1. Không khớp mẫu của brandname mượn "Baotrixemay" → eSMS trả
///      `CodeResult 146 — Sai template`, không tin nào ra khỏi hệ thống.
///   2. Viết CÓ DẤU và có dấu gạch ngang `–` (U+2013) → bị đẩy sang UCS-2,
///      70 ký tự/đoạn thay vì 160, tiền tin nhắn tăng gấp bội. Đây đúng là cái
///      bẫy đã ghi ở `docs/SMS-HOA-DON.md` mục 6.1 mà chính chỗ này vẫn dính.
///
/// Tham số `_house` giữ lại cho khỏi phải sửa nơi gọi; mẫu brandname không có
/// chỗ nhét thông tin ngân hàng — người thuê xem trong ảnh hoá đơn mở từ link.
export function buildInvoiceSmsMessage(
  invoice: {
    room_nos: string[];
    period_start: string;
    total_amount: number;
    due_date: string;
    public_code?: string | null;
  },
  _house?: unknown,
): string {
  const ky = invoice.period_start.slice(5, 7) + "/" +
    invoice.period_start.slice(0, 4);
  const host = (Deno.env.get("SUPABASE_URL") ?? "").replace(/^https?:\/\//, "");
  const link = invoice.public_code
    ? `${host}/functions/v1/invoice/${invoice.public_code}`
    : host;
  const phong = (invoice.room_nos ?? []).join(",");
  const tien = formatVndNumber(invoice.total_amount);
  const han = formatDdMm(invoice.due_date);
  let tomTat = `BizTown P${phong} ${tien}d han ${han}`;
  if (tomTat.length > 50) tomTat = `BizTown ${tien}d han ${han}`;

  return `Quy khach da den thoi gian bao tri lan ${ky} xe ${link}. ` +
    `Vui long lien he ${tomTat} de duoc huong dan. Tran trong.`;
}
